### DISCUSSION TITLE
aarch64 kylin CMAKE_STATICBUILD=1 tools/staticbuild/build.sh cpu > build.log 2>&1

### DISCUSSION THREAD

**Discussion by Duanexiao** [apache/incubator-mxnet#20414] [Q&A]
os-release info:

NAME="Kylin Linux Advanced Server"
VERSION="V10 (Tercel)"
ID="kylin"
VERSION_ID="V10"
PRETTY_NAME="Kylin Linux Advanced Server V10 (Tercel)"
ANSI_COLOR="0;31"

incubator-mxnet branch: 1.9.x

execute command: CMAKE_STATICBUILD=1 tools/staticbuild/build.sh cpu > build.log 2>&1

error info:
/usr/bin/ld: ../staticdeps/lib64/libpng.a(pngrutil.c.o): in function `png_inflate_claim': pngrutil.c:(.text+0xff8): undefined reference to `inflateValidate'

**Comment by Duanexiao** [comment]
[build.log](https://github.com/apache/incubator-mxnet/files/6754209/build.log)

**Comment by Duanexiao** [comment]
for build log detail info,you can download the build.log file above

**Reply by Duanexiao** [reply]
google the error info,returns me the answer to install zlib.Exactly, the system already exists zlib with 1.2.11,and you can find the zlib with 1.2.6 in the staticdeps.I have no idea to continue solve the problem,so sos for help.Many tks!

**Comment by leezu** [comment]
Please follow the build from source instructions https://mxnet.apache.org/get_started/build_from_source The `tools/staticbuild/build.sh` is a special script used to statically link dependencies for publication on https://pypi.org/project/mxnet. Do you need that?

**Reply by Duanexiao** [reply]
Thank you for you reply.I want to build mxnet for python on aarch64.When I pip install mxnet, the libmxnet.so file is not avaliable.So I need to build the libmxnet.so

**Reply by leezu** [reply]
Great. For that you should *not* use `tools/staticbuild/build.sh` but rather follow the instructions in https://mxnet.apache.org/get_started/build_from_source (which also links to https://github.com/apache/incubator-mxnet/blob/9ebe801e14dd657ee9f410f934b9e2d94416ea29/config/linux.cmake#L18-L36)

**Reply by Duanexiao** [reply]
Thank you.I am newbee for this operation.So I am not sure about your answer.DId you mean that: 
cd mxnet/config;mkdir build; cd build;cmake ..;cmake --build .
I did like this,but there is no cmakelist.txt.I should copy the mxnet/CMakeList.txt here?Looking for detail build operation on the get_started.

**Reply by leezu** [reply]
No, you shouldn't do `cd mxnet/config`. Just follow the commands listed above in sequence from the mxnet root directory. For example:

```
cd /data/mxnet
cp config/linux.cmake config.cmake 
mkdir build; cd build 
cmake .. 
cmake --build . 
```

**Comment by Duanexiao** [comment]
cmake --build . error:
/data/mxnet/3rdparty/onednn/src/cpu/gemm_convolution.cpp: In lambda function:
/data/mxnet/3rdparty/onednn/src/cpu/gemm_convolution.cpp:321:29: sorry, unimplemented: unexpected AST of kind omp_simd
                             });
                             ^
/data/mxnet/3rdparty/onednn/src/cpu/gemm_convolution.cpp:321: confused by earlier errors, bailing out
gmake[2]: *** [3rdparty/onednn/src/cpu/CMakeFiles/dnnl_cpu.dir/build.make:336: 3rdparty/onednn/src/cpu/CMakeFiles/dnnl_cpu.dir/gemm_convolution.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:1338: 3rdparty/onednn/src/cpu/CMakeFiles/dnnl_cpu.dir/all] Error 2
gmake: *** [Makefile:141: all] Error 2

gcc version: 9.4.0
cmake version:3.15.0

**Comment by Duanexiao** [comment]
I've solved this problem.So here I want to share my answer.Below as my whole reproduciable operation.All my operations in docker
arch: arm64
os: centos7

step1: 

centos7 arm64 dockerfile:
`FROM centos:centos7`
`ENV TZ=Asia/Shanghai`
`RUN yum install kde-l10n-Chinese -y && \`
`yum install glibc-common -y && \`
`yum install cronie -y && \`
`localedef -c -f UTF-8 -i zh_CN zh_CN.utf8 && \`
`export LANG=zh_CN.UTF-8 && \`
`echo "export LANG=zh_CN.UTF-8" >> /etc/locale.conf && \`
`ln -snf /usr/share/zoneinfo/$TZ /etc/localtime && echo $TZ > /etc/timezone && \`
`yum clean all`
`ENV LANG zh_CN.UTF-8`
`ENV LC_ALL zh_CN.UTF-8`
`ENTRYPOINT /bin/bash`

use this dockerfile to build image, command is below:
`docker build -t shakespeare-base-centos7-arm:0.1 .`

bash into one container based on docker image above
`docker run -it --entrypoint /bin/bash 6cb7b6f6cd18`

step 2:

install high version gcc
`yum install centos-release-scl -y`
`yum install devtoolset-8 -y`
#activate devtoolset-8 env
`source /opt/rh/devtoolset-8/enable`

step 3: 

install high version cmake
`rpm -ivh https://mirrors.tuna.tsinghua.edu.cn/epel/epel-release-latest-7.noarch.rpm`
`yum install cmake3 -y`
`ln -s /usr/bin/cmake3 /usr/bin/make`

step 4: 

source code compile high version opencv
`mkdir ~/opencv_build && cd ~/opencv_build`
`git clone https://github.com/opencv/opencv.git`
`git clone https://github.com/opencv/opencv_contrib.git`
`cd ~/opencv_build/opencv && make build && cd build`
`cmake3 -D CMAKE_BUILD_TYPE=RELEASE `
   ` -D CMAKE_INSTALL_PREFIX=/usr/local `
   ` -D INSTALL_C_EXAMPLES=ON `
    `-D INSTALL_PYTHON_EXAMPLES=ON `
    `-D OPENCV_GENERATE_PKGCONFIG=ON `
    `-D OPENCV_EXTRA_MODULES_PATH=~/opencv_build/opencv_contrib/modules `
  `  -D BUILD_EXAMPLES=ON ..`

#notice that successfully installed display like these：
#-- Configuring done
#-- Generating done
#-- Build files have been written to: /root/opencv_build/opencv/build
 
 
#next step
`make -j $(cat /proc/cpuinfo | grep processor | wc -l)`
 
 
#next step
`make install`
 
 
#soft link
`ln -s /usr/local/lib64/pkgconfig/opencv4.pc /usr/share/pkgconfig/opencv4.pc`
`ldconfig`
 
 
#check opencv version
`pkg-config --modversion opencv4`

step 5: 

source code compile openblas

`git clone https://github.com/xianyi/OpenBLAS.git`
`cd OpenBLAS`
`make -j $(cat /proc/cpuinfo | grep processor | wc -l)`
#make successfully display like this：
make[1]: Leaving directory '/root/OpenBLAS/exports'
 
 OpenBLAS build complete. (BLAS CBLAS LAPACK LAPACKE)
 
  OS               ... Linux
  Architecture     ... arm64
  BINARY           ... 64bit
  C compiler       ... GCC  (cmd & version : cc (GCC) 8.3.1 20190311 (Red Hat 8.3.1-3))
  Fortran compiler ... GFORTRAN  (cmd & version : GNU Fortran (GCC) 8.3.1 20190311 (Red Hat 8.3.1-3))
  Library Name     ... libopenblas_tsv110p-r0.3.15.dev.a (Multi-threading; Max num-threads is 128)
 
To install the library, you can run "make PREFIX=/path/to/your/installation install".
 
#make install set prefix so that you donot need to change findOpenBLAS file content
`make PREFIX=/opt/OpenBLAS install`

step 6: 

source code compile mxnet

`git clone --recursive https://github.com/apache/incubator-mxnet mxnet`
`cd mxnet`
`mkdir build`
`cp config/linux.cmake config.cmake`
`cd build`
`cmake ..`
#cmake ok display like these: 
-- Using unsigned short
-- Check if the system is big endian - little endian
-- /root/mxnet/3rdparty/dmlc-core/cmake/build_config.h.in -> include/dmlc/build_config.h
-- Performing Test SUPPORT_MSSE2
-- Performing Test SUPPORT_MSSE2 - Failed
-- Using 64-bit integer for tensor size
-- Found Python3: /usr/bin/python3.6 (found version "3.6.8") found components:  Interpreter
-- Configuring done
-- Generating done
-- Build files have been written to: /root/mxnet/build
 
#next step
`cmake --build .`
 
 #cmake build ok,you'll find libmxnet.so file

step 7:

python install mxnet

#after compile, we should change directory to root of mxnet !!!
#download Archiconda3-0.2.3-Linux-aarch64.sh file and install that, you should  pip install numpy opencv and many other related modules

`mkdir /opt/wuzhi_ai`
`tar -xf anaconda3_aarch64.tar.gz -C /opt/wuzhi_ai`
`echo "export PATH=/opt/wuzhi_ai/anaconda3/bin:$PATH" >> /etc/profile`
`source /etc/profile`
#last steps are install python environment, you can change these operation as you like
`python -m pip install --user -e ./python`


That's all
